### DISCUSSION TITLE
Pull request conflicts: Support `merge=union` in .gitattributes file

### DISCUSSION THREAD

**Discussion by rtzoeller** [community/community#9288] [Pull Requests]
This is a continuation of https://github.com/isaacs/github/issues/487, as I could not find an existing one.

https://github.com/nix-rust/nix has recently seen a number of PRs which only conflict on a CHANGELOG.md file. This forces PRs to frequently need to be manually rebased. The changelog is marked as `merge=union` in a `.gitattributes` file so these rebases are trivial to perform locally, but it can make accepting contributions from external contributors challenging.

It would be helpful if GitHub supported this attribute natively, so manual rebases were not required.

**Comment by pscott31** [comment]
Please GitHub :)

**Comment by niuhuii** [comment]
hope this

**Comment by matthiasgoergens** [comment]
Looks like github is using a bad re-implementation of half of git internally?

**Comment by kit-ty-kate** [comment]
\*taps the mic\* is anyone even working on the core product at Github anymore?

**Reply by jaimeagudo** [reply]
Bill Gates 🧌

**Comment by fboja** [comment]
+1 for getting at least partial support for `merge=union` being respected from `.gitattributes` .

We have a whole team running into merge conflicts on PRs daily due this issue.

We can avoid the conflicts by doing client-side branch merging (as that respects `.gitattributes` merge rules), but that would require relaxing branch protection rules to allow commits directly on target branches which would be a major downgrade for keeping our codebase safe.

TL;DR: GitHub PRs ignoring `.gitattributes` forces us to manually having to work around basic Git functionality missing from Github.

**Comment by Potherca** [comment]
For the sake of completion (from the O.G. Isaacs issue tracker):

> gitlab [now has merge=union support](https://gitlab.com/gitlab-org/gitlab-ce/issues/17325), because the underlying [rugged](https://github.com/libgit2/rugged) library has the support since v0.25 (use v0.25.1.1 with critical fixes) and as it's github's library, i believe they rely on that library as well. so perhaps it just works now if they upgraded library in their backends?

_Originally posted by @glensc in https://github.com/isaacs/github/issues/487#issuecomment-324538695 on Aug 24, 2017_

and

> libgit2 has added support for merge=union (or custom merge) since 2016 (https://github.com/libgit2/libgit2/pull/3564) and git-merge-changelog custom driver since http://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=blob;f=lib/git-merge-changelog.c

_Originally posted by @timotheecour in https://github.com/isaacs/github/issues/487#issuecomment-619449802 on Apr 26, 2020_

Just so you know...

**Comment by cadesalaberry** [comment]
I found this [message](https://github.com/olivierlacan/keep-a-changelog/issues/56#issuecomment-350291283) from github support, dated circa Dec 8, 2017 that states that github uses a custom `.gitattributes` file which cannot be updated:
```
Hi AJ,

Thanks for reaching out to GitHub Support!

I'm afraid what you're seeing is currently expected behavior.

GitHub doesn't consider user-defined .gitattributes files (normally, we use our own .gitattrbutes file which you can't change). I agree it would be useful if you could define that yourself and I'll pass your feedback to the team, but I can't promise when such a feature might be available.

The only workaround I could suggest is to merge pull requests in your local clone (and not via the web UI), by using the instructions shown in the merge box for each pull request. It's not a very elegant solution, I agree, but something to consider if this is causing you problems.

Thanks for the feedback here, we really appreciate it. Please feel free to let us know if there is anything else we can help with!
```

We have been running into this merge driver issue with our `translations_fr-FR.json` file. We were pretty disappointed when we realised github did not support this feature. We are using [Codemod](https://codemod.com/) to progressively bring i18n to our website. This means adding text to the `translations_fr-FR.json` file in EVERY PR. 90% of the time, merging a PR conflicts with the other that have been prepared, forcing us to re-generate or rebase. For the ~400 PR that we should merge, this is a big pain point.

In case it helps someone else, to go around this issue, we ended up using a "compile" script to gather all the translations in one file with the following architecture:
```
assets/i18n/codmod/
+- extractions/
|  +- pr-2350.fr-FR.json
|  +- pr-850.fr-FR.json
+- codmod_fr-FR.json <= merged version of all the json present in the extractions folder, manually created with a script before build
+- codmod_en-US.json <= translated version of the source fr-FR, completed by Crowdin
```

**Comment by stopiccot-tds** [comment]
Please add this feature

**Comment by dtaivpp** [comment]
Yeah someone at GitHub needs to pay attention to this. Wait..

**Comment by dlim-yc** [comment]
Please add this feature!

**Comment by jaimeagudo** [comment]
Wake up, Microsoft listening to users???: I just

 `git cherry-pick  <commit-hash>`
